Source/NSSwitch.m

```diff
@@ -0,0 +1,291 @@
+{
+  _state = s;
+  [self setNeedsDisplay];
+}
+
+- (NSControlStateValue) state
+{
+  return _state;
+}
+
+- (void) setAction: (SEL)action
+{
+  _action = action;
+}
+
+- (SEL) action
+{
+  return _action;
+}
+
+- (void) setTarget: (id)target
+{
+  _target = target;
+}
+
+- (id) target
+{
+  return _target;
+}
+
+- (void) setEnabled: (BOOL)flag
+{
+  _enabled = flag;
+  [self setNeedsDisplay];
+}
+
+- (BOOL) isEnabled
+{
+  return _enabled;
+}
+
+- (void) setDoubleValue: (double)val
+{
+  if (val < 1.0)
+    {
+      [self setState: NSControlStateValueOff];
+    }
+  else
+    {
+      [self setState: NSControlStateValueOn];
+    }
+}
+
+- (double) doubleValue
+{
+  return (double)(([self state] == NSControlStateValueOn) ? 1.0 : 0.0);
+}
+
+- (void) setFloatValue: (float)val
+{
+  [self setDoubleValue: (double)val];
+}
+
+- (float) floatValue
+{
+  return (float)[self doubleValue];
+}
+
+- (void) setIntValue: (int)val
+{
+  [self setDoubleValue: (double)val];
+}
+
+- (int) intValue
+{
+  return (int)[self doubleValue];
+}
+
+- (void) setIntegerValue: (NSInteger)val
+{
+  [self setDoubleValue: (double)val];
+}
+
+- (NSInteger) integerValue
+{
+  return (NSInteger)[self doubleValue];
+}
+
+- (void) setStringValue: (NSString *)val
+{
+  [self setDoubleValue: [val doubleValue]];
+}
+
+- (NSString *) stringValue
+{
+  return [NSString stringWithFormat: @"%ld", [self integerValue]];
+}
+
+- (void) setObjectValue: (id)obj
+{
+  if ([obj respondsToSelector: @selector(stringValue)])
+    {
+      [self setStringValue: [obj stringValue]];
+    }
+}
+
+- (id) objectValue
+{
+  return [self stringValue];
+}
+
+- (void) drawRect: (NSRect)rect
+{
+  [[GSTheme theme] drawSwitchInRect: rect
+                           forState: _state
+                            enabled: [self isEnabled]];
+}
+
+- (void) mouseDown: (NSEvent *)event
+{
+  if (![self isEnabled])
+    {
+      [super mouseDown: event];
+      return;
+    }
+  
+  if (_state == NSControlStateValueOn)
+    {
+      [self setState: NSControlStateValueOff];
+    }
+  else
+    {
+      [self setState: NSControlStateValueOn];
+    }
+
+  if (_action)
+    {
+      [self sendAction: _action
+                    to: _target];
+    }
+}
+
+// Accessibility
+- (NSRect)accessibilityFrame
+{
+  return [self frame];
+}
+
+- (NSString *)accessibilityIdentifier
+{
+  return nil;
+}
+
+- (id)accessibilityParent
+{
+  return nil;
+}
+
+- (BOOL)isAccessibilityFocused
+{
+  return NO;
+}
+
+- (NSString *)accessibilityLabel
+{
+  return nil;
+}
+
+- (BOOL)accessibilityPerformPress
+{
+  return NO;
+}
+
+- (BOOL) accessibilityPerformDecrement
+{
+  return NO;
+}
+
+- (BOOL) accessibilityPerformIncrement
+{
+  return NO;
+}
+
+- (NSString *) accessibilityValue
+{
+  return nil;
+}
+
+// NSCoding
+- (id) initWithCoder: (NSCoder *)coder
+{
+  if ((self = [super initWithCoder: coder]) != nil)
+    {
+      if ([coder allowsKeyedCoding])
+        {
+          if ([coder containsValueForKey: @"NSControlContents"])
```

Comment:
My feeling is that those should be moved to NSControl but for now it is OK. It looks like the Cocoa classes have changed their keyed encoding over the years.